### PR TITLE
Feature/article favorites count

### PR DESCRIPTION
**Pull request by spinder**
+ New Feature Details: https://github.com/spinder/kotlin-ktor-realworld-example-app/issues/1
+ Tests included
+ Updateed Github Actions included
+ See details of PR implementation here:
https://github.com/spinder/kotlin-ktor-realworld-example-app/issues/1

### CODE REVIEW

File: src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    init {
+        transaction {
+            SchemaUtils.create(Articles, ArticleTags, Favorites)
+        }
+    }
+
+    private fun toArticle(row: ResultRow, currentUserId: Long?): Article {
+        val articleId = row[Articles.id]
+        val authorRow = Users.selectAll().firstOrNull { it[Users.id].value == row[Articles.authorId] }
+        val author = authorRow?.let { Users.toDomain(it) }
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt:50]: **[bug]** `toArticle` sets `author` from `Users.toDomain()`, which includes `id`, `email`, and `password`. Jackson will serialize that nested `User` on every article response (`GET /api/articles`, feed, popular, get-by-slug, favorite). RealWorld requires an author profile (`username`, `bio`, `image`, `following`) and must not return credentials. `CommentRepository.toComment` has the same leak.

**Suggestion:** Map authors to the `Profile` shape (no email/password/id) and populate `following` from the current user. Add a test that article JSON has no `password` or `email` under `author`.

File: src/main/kotlin/io/realworld/app/web/controllers/ArticleController.kt [hunk 2]
```diff
@@ -1,77 +1,88 @@
...
 
-    fun favorite(ctx: ApplicationCall): ArticleDTO {
-        ctx.parameters["slug"]
-        //            articleService.favorite(ctx.attribute("email"), slug).apply {
-//                ctx.json(ArticleDTO(this))
-//            }
-        return ArticleDTO(null)
+    suspend fun delete(ctx: ApplicationCall) {
+        val slug = ctx.parameters["slug"]!!
+        articleService.delete(slug)
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/web/controllers/ArticleController.kt:69]: **[bug]** `delete` calls `articleService.delete(slug)` with no check that the authenticated user is the author. Any logged-in user can delete any article. `update` (line 62) has the same gap: it never compares `currentUser` to `article.authorId`.

**Suggestion:** Load the article, return 403 (or 404) when `authorId != currentUser.id`, and add a test where a second user cannot update or delete someone else's article.

File: src/main/kotlin/io/realworld/app/web/controllers/CommentController.kt [hunk 3]
```diff
@@ -1,32 +1,35 @@
...
     }
 
-    fun delete(ctx: ApplicationCall) {
-        val slug = ctx.parameters["slug"]
-        val id = ctx.parameters["id"]
-//        commentService.delete(id, slug)
+    suspend fun delete(ctx: ApplicationCall) {
+        val slug = ctx.parameters["slug"]!!
+        val id = ctx.parameters["id"]!!.toLong()
+        commentService.delete(id, slug)
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/web/controllers/CommentController.kt:32]: **[bug]** `delete` only needs a valid JWT; it never checks comment author (or article author). `CommentService.delete` / `CommentRepository.delete` just `deleteWhere` by id+slug, so any authenticated user can remove any comment. Missing comments still return 200.

**Suggestion:** Require the comment author (RealWorld behavior), return 404 when the comment is missing, and add a cross-user delete test.

File: src/main/kotlin/io/realworld/app/web/controllers/ArticleController.kt [hunk 4]
```diff
@@ -1,77 +1,88 @@
...
-//        articleService.findFeed(ctx.attribute("email"), limit.toInt(), offset.toInt()).also { articles ->
-//            ctx.json(ArticlesDTO(articles, articles.size))
-//        }
-        return ArticlesDTO(listOf(), 1)
+    suspend fun popular(ctx: ApplicationCall) {
+        val limit = ctx.request.queryParameters["limit"]?.toIntOrNull() ?: 20
+        val offset = ctx.request.queryParameters["offset"]?.toIntOrNull() ?: 0
+        val currentUser = ctx.authentication.principal<User>()
+        val articles = articleService.findPopular(currentUser?.id, limit, offset)
+        ctx.respond(ArticlesDTO(articles, articles.size))
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/web/controllers/ArticleController.kt:32]: **[bug]** `popular` (and `findBy`/`feed`) respond with `ArticlesDTO(articles, articles.size)`, so `articlesCount` equals the current page length. RealWorld clients use `articlesCount` as the total matching rows to compute pagination (`ceil(articlesCount / limit)`). The new test at `PopularArticleControllerTest.kt:99` locks this in (`limit=2` with 3 articles expects `articlesCount == 2`). OpenAPI even documents the wrong meaning.

**Suggestion:** Count the full filtered set (or `COUNT(*)` before `LIMIT`/`OFFSET`) and return that as `articlesCount`. Keep `articles.length` as the page size only. Fix the popular test to expect `3` when `limit=2`.

**Review by obey-agent — COMMENTED** [review]
## Summary

This PR restores the previously stubbed RealWorld article/comment/profile stack, prefixes routes with `/api`, and adds `GET /api/articles/feed/popular`. The wiring is enough for the existing JUnit suite to go green, but several new code paths break the RealWorld contract or leak data: article `author` is a full `User` (including stored password), mutate endpoints skip ownership checks, `articlesCount` is the page length rather than the total, and list/feed results are not newest-first. Popular-feed tests encode those contract mistakes and do not cover offset, multi-level sort, or authorization.

## Issue counts by severity

- bugs: 8
- suggestions: 2
- nits: 0

File: src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                val favUser = Users.select { Users.username eq favorited }.firstOrNull()
+                if (favUser != null) {
+                    val favUserId = favUser[Users.id].value
+                    val favIds = Favorites.select { Favorites.userId eq favUserId }.map { it[Favorites.articleId] }.toSet()
+                    filtered = filtered.filter { (id, _, _) -> id in favIds }
+                } else {
+                    filtered = listOf()
+                }
+            }
+            filtered.drop(offset).take(limit).map { (_, _, row) -> toArticle(row, currentUserId) }
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt:122]: **[bug]** `findBy` applies filters then `drop(offset).take(limit)` with no `ORDER BY createdAt DESC`. `findFeed` (line 147) does the same. RealWorld list/feed must return most recently created articles first; H2 `selectAll()` is insertion/PK order (oldest first).

**Suggestion:** Sort by `Articles.createdAt` descending (and a stable tie-breaker) before pagination, ideally in SQL rather than in memory.

File: src/main/kotlin/io/realworld/app/domain/service/ArticleService.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    fun findFeed(currentUserId: Long?, limit: Int, offset: Int): List<Article> =
+        articleRepository.findFeed(currentUserId, limit, offset)
+
+    fun findBySlug(slug: String, currentUserId: Long? = null): Article? =
+        articleRepository.findBySlug(slug, currentUserId)
+
+    fun create(email: String, article: Article): Article {
+        val user = userRepository.findByEmail(email) ?: error("User not found: $email")
+        val slug = article.title!!.toSlug()
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/domain/service/ArticleService.kt:25]: **[bug]** Slug is `title.toSlug()` with no uniqueness suffix. `Articles.slug` is uniquely indexed, so a second article with the same title fails the insert and the global `StatusPages` handler turns it into HTTP 500. `update` also never rewrites the slug when the title changes, which the RealWorld spec requires.

**Suggestion:** Append a unique suffix (id or random) on collision, regenerate slug on title change, and add tests for duplicate titles and title updates.

File: src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                .map { toArticle(it, currentUserId) }
+        }
+    }
+
+    fun update(slug: String, article: Article, currentUserId: Long?): Article? {
+        return transaction {
+            Articles.update({ Articles.slug eq slug }) { row ->
+                if (article.title != null) row[Articles.title] = article.title
+                if (article.description != null) row[Articles.description] = article.description
+                row[Articles.body] = article.body
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt:156]: **[bug]** `Articles.update` always assigns `row[Articles.body] = article.body` even when the client omitted body. `title`/`description` are guarded with null checks; body is not. Combined with `Article.body` being a non-null `String`, a partial update can wipe the body (empty string) or force clients to resend it.

**Suggestion:** Treat body like the other optional fields (`if (article.body.isNotBlank())` or make `body` nullable on the update DTO) and skip the column when it was not provided.

File: src/main/kotlin/io/realworld/app/web/controllers/ProfileController.kt [hunk 8]
```diff
@@ -1,24 +1,29 @@
...
-    //class ProfileController(private val userService: UserService) {
-    fun get(ctx: ApplicationCall) {
-        ctx.parameters["username"]
-//            userService.getProfileByUsername(ctx.attribute("email")!!, usernameFollowing).also { profile ->
-//                ctx.json(ProfileDTO(profile))
+class ProfileController(private val userService: UserService) {
+    suspend fun get(ctx: ApplicationCall) {
+        val username = ctx.parameters["username"]!!
+        val email = ctx.authentication.principal<User>()?.email ?: ""
+        ctx.respond(ProfileDTO(userService.getProfileByUsername(email, username)))
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/web/controllers/ProfileController.kt:15]: **[bug]** GET profile now calls `userService.getProfileByUsername`, which uses `UserRepository.findIsFollowUser`. That helper joins `Follows.user eq Users.id` (current user as the *followed* row) and `Follows.follower eq userIdToFollow` (profile user as follower), so `following` is inverted: it is true when the profile follows the caller, not when the caller follows the profile. POST follow hides this by hardcoding `following=true`; tests never GET after follow.

**Suggestion:** Query `Follows.user eq profileUserId AND Follows.follower eq currentUserId`. Add a test that GET `/api/profiles/{username}` returns `following=true` after a successful follow.

File: src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+                    filtered = listOf()
+                }
+            }
+            filtered.drop(offset).take(limit).map { (_, _, row) -> toArticle(row, currentUserId) }
+        }
+    }
+
+    fun findPopular(currentUserId: Long?, limit: Int, offset: Int): List<Article> {
+        return transaction {
+            Articles.selectAll().toList()
```
Comment (obey-agent) [inline src/main/kotlin/io/realworld/app/domain/repository/ArticleRepository.kt:128]: **[suggestion]** `findPopular` loads every article, then runs a `Favorites.count()` per row, then `toArticle` counts favorites/tags/author again. This is O(n) queries and O(n) memory; pagination happens after the full sort. Negative `limit`/`offset` from the controller (`toIntOrNull() ?: 20`) also throw `IllegalArgumentException` from `drop`/`take` and become HTTP 500.

**Suggestion:** Use a grouped SQL count (`LEFT JOIN Favorites GROUP BY article ORDER BY count DESC LIMIT/OFFSET`). Reject or clamp non-negative `limit`/`offset` in the controller.

File: src/test/kotlin/io/realworld/app/web/controllers/PopularArticleControllerTest.kt [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                "/api/articles",
+                ArticleDTO(Article(title = "Article $i", description = "desc", body = "body", tagList = listOf("tag$i")))
+            )
+        }
+
+        val response = appRule.http.get<ArticlesDTO>("/api/articles/feed/popular?limit=2")
+
+        assertEquals(HttpStatus.SC_OK, response.status)
+        assertEquals(2, response.body.articles.size)
+        assertEquals(2, response.body.articlesCount)
```
Comment (obey-agent) [inline src/test/kotlin/io/realworld/app/web/controllers/PopularArticleControllerTest.kt:99]: **[suggestion]** The new tests miss the cases that would catch the bugs above: they assert `articlesCount == page size`, never pass `offset`, never use two non-zero favorite counts, never check unauthenticated `favorited=false` vs authenticated `favorited=true`, and there are no tests that a non-author cannot delete/update an article or comment. CI (`./gradlew test`) will gate on this suite, so these gaps mean the pipeline will stay green while the API is wrong.

**Suggestion:** Add offset + total-count assertions, a 3-way favorites sort, favorited-flag coverage with and without a token, and ownership 403 tests. Do not assert `articlesCount == articles.size` when a limit is applied.
